### PR TITLE
Add in some PhaseII Code.

### PR DESCRIPTION
**Pull request by trmazi**
Not all of this code is working! GFDM V8 does, and Hello! Pop'n does. they're both VERY messy but they should be able to merge.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -15,3 +15,4 @@ bemani.egg-info/
 server/
 emblems
 .vscode/
+.DS_Store
```
Comment (DragonMinded) [inline .gitignore:18]: https://i.imgur.com/t9Kkt8h.jpg

File: bemani/backend/dm/base.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+from bemani.backend.base import Base
+from bemani.backend.core import CoreHandler, CardManagerHandler
+from bemani.common import ValidatedDict, GameConstants, Time, DBConstants, Parallel, Model
+from bemani.data import UserID, Score, Data
+from bemani.protocol import Node
+
+class DrummaniaBase(CoreHandler, CardManagerHandler, Base):
+    """
+    Trying to work on GFDM right now. starting off strong with DM :)
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:11]: Please remove irrelevant comments, they don't belong as docstrings. If you want to instead document what is expected to work and what isn't, this is an okay place to do that.

File: bemani/backend/dm/base.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    GRADE_S = DBConstants.DRUMMANIA_GRADE_S
+
+    CHART_TYPE_BEGINNER = 0
+    CHART_TYPE_BASIC = 1
+    CHART_TYPE_ADVANCED = 2
+    CHART_TYPE_EXTREME = 3
+
+    def __init__(self, data: Data, config: Dict[str, Any], model: Model) -> None:
+        super().__init__(data, config, model)
+        if model.rev == 'X':
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:37]: Does DM have omnimix or is this copypasta?

File: bemani/backend/dm/base.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+        for songid in remote_attempts:
+            if songid not in attempts:
+                attempts[songid] = {}
+
+            for songchart in remote_attempts[songid]:
+                if songchart not in attempts[songid]:
+                    attempts[songid][songchart] = {
+                        'total': 0,
+                        'clears': 0,
+                        'average': 0,
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:187]: Do we not update the average anywhere after here? Is this copied from some buggy code elsewhere?

File: bemani/backend/dm/base.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+            highscore = score >= oldscore.points
+            score = max(oldscore.points, score)
+            scoredata = oldscore.data
+
+        # If we have a combo, replace it
+        scoredata.replace_int('combo', max(scoredata.get_int('combo'), combo))
+        history.replace_int('combo', combo)
+
+        #write our skill points
+        print(skill_perc)
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:266]: Please go through and remove all debug prints from the code before it can be merged.

File: bemani/backend/dm/base.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+            raised = score > oldscore.points
+            highscore = score >= oldscore.points
+            score = max(oldscore.points, score)
+            scoredata = oldscore.data
+
+        # If we have a combo, replace it
+        scoredata.replace_int('combo', max(scoredata.get_int('combo'), combo))
+        history.replace_int('combo', combo)
+
+        #write our skill points
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:265]: Fairly sure this will result in a lint error for no space between # and write. Please make sure before you submit in the future that you run `./verifylint` as well as `./veritytyping` to catch these sorts of things.

File: bemani/backend/dm/base.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+        #write our skill percent
+        scoredata.replace_int('skill_perc', max(scoredata.get_int('skill_perc'), skill_perc))
+        history.replace_int('skill_perc', skill_perc)
+
+        #did we clear the song?
+        scoredata.replace_int('clear', max(scoredata.get_int('clear'), clear))
+        history.replace_int('clear', clear)
+
+        #did we cheat?
+        #scoredata.replace_int('autoclear', max(scoredata.get_int('autoclear'), autoclear))
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:280]: Shouldn't this function just reject the score outright if its a cheat? Why is this here?

File: bemani/backend/dm/base.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+
+        # GFDM for all versions sends all of the songs
+        # a player played at the end of the round. It doesn't send timestamps
+        # for those songs (Jubeat does). So, if a user plays the same song/chart
+        # more than once in a round, we will end up failing to store the attempt
+        # since we don't allow two of the same attempt at the same time for the
+        # same user and song/chart. So, bump the timestamp by one second and retry
+        # well past the maximum number of songs.
+        now = Time.now()
+        for bump in range(10):
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:315]: This just unconditionally adds 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 + 9 + 10 to the timestamp. You aren't checking for anything here?

File: bemani/backend/dm/base.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+            timestamp = now + bump
+
+        print(songid)
+        
+
+        # Write the new score back
+        self.data.local.music.put_score(
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:322]: If you need to modify the timestamp, it should be done in a try/catch loop with this function which will fail when inserting a duplicate score. Was this tested? I presume not because I don't think you wrote any traffic tests for this.

File: bemani/backend/dm/dmv4.py [hunk 10]
```diff
@@ -0,0 +1,521 @@
...
+
+class DrummaniaV4(
+    EventLogHandler,
+    DrummaniaBase,
+):
+
+    name = "Drummania V4"
+    version = VersionConstants.DRUMMANIA_V4
+
+    GAME_LIMITED_LOCKED = 1
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv4.py]: This entire file looks like a rough copy-pasta of SDVX, including packets and blocks and limited stuff. I doubt DMV4 works the same way. Honestly if you need to check in stubs, this should be an empty file. Preferrably it shouldn't even be included in the review. I'm going to skip reviewing this whole file because it looks like its not mergeable as-is.

File: bemani/backend/dm/dmv5.py [hunk 11]
```diff
@@ -0,0 +1,521 @@
...
+
+class DrummaniaV5(
+    EventLogHandler,
+    DrummaniaBase,
+):
+
+    name = "Drummania V5"
+    version = VersionConstants.DRUMMANIA_V5
+
+    GAME_LIMITED_LOCKED = 1
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv5.py]: Same comment here, get rid of completely useless code that just copies SDVX please.

File: bemani/backend/dm/dmv6.py [hunk 12]
```diff
@@ -0,0 +1,521 @@
...
+
+    @classmethod
+    def get_settings(cls) -> Dict[str, Any]:
+        """
+        Return all of our front-end modifiably settings.
+        """
+        return {
+            'bools': [
+                {
+                    'name': 'Disable Online Matching',
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv6.py]: Same comment here... drop this file.

File: bemani/backend/dm/dmv7.py [hunk 13]
```diff
@@ -0,0 +1,2512 @@
...
+    DrummaniaBase,
+):
+
+    name = "Drummania V7"
+    version = VersionConstants.DRUMMANIA_V7
+
+    GAME_LIMITED_LOCKED = 1
+    GAME_LIMITED_UNLOCKED = 2
+
+    GAME_CURRENCY_PACKETS = 0
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: And here...

File: bemani/backend/dm/dmv7.py [hunk 14]
```diff
@@ -0,0 +1,2512 @@
...
+
+        return root
+
+    def handle_cardutil_regist_request(self, request: Node) -> Node:
+
+        data = request.child('data')
+
+        refid = data.child_value('refid')
+        name = data.child_value('name')
+        gdp = 0
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: Why is all this here instead of just in the dictionary below? No need for so many tempvars.

File: bemani/backend/dm/dmv7.py [hunk 15]
```diff
@@ -0,0 +1,2512 @@
...
+
+        root = Node.void('gametop')
+
+        playergt = request.child('player')
+        refid = playergt.child_value('refid')
+
+        userid = self.data.remote.user.from_refid(self.game, self.version, refid)
+        profile = self.get_profile(userid)
+        
+
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: W
H
Y
?

File: bemani/backend/dm/dmv7.py [hunk 16]
```diff
@@ -0,0 +1,2512 @@
...
+        battledata.add_child(Node.u32('perfect', profile.get_int('perfect', 0)))
+        battledata.add_child(Node.u32('great', profile.get_int('great', 0)))
+        battledata.add_child(Node.u32('good', profile.get_int('good', 0)))
+        battledata.add_child(Node.u32('poor', profile.get_int('poor', 0)))
+        battledata.add_child(Node.u32('miss', profile.get_int('miss', 0)))
+
+        history = Node.void('history')
+        battledata.add_child(history)
+
+        round_before = Node.void('round')
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: Use a loop for all this, why is it copy-pasta'd 10 times?

File: bemani/backend/dm/dmv7.py [hunk 17]
```diff
@@ -0,0 +1,2512 @@
...
+            [
+                0,
+                0,
+            ]
+        )
+        #info.add_child(champ_result)
+
+
+
+        #no idea what this does lol
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: What are these comments?

File: bemani/backend/dm/dmv7.py [hunk 18]
```diff
@@ -0,0 +1,2512 @@
...
+        player = Node.void('player')
+        player.set_attribute('no', "1")
+
+        root.add_child(player)
+
+        player.add_child(Node.u8('player_type', 0))
+        player.add_child(Node.string('my_rival_id', "0"))
+        player.add_child(Node.u8('mode', profile.get_int('playmode')))
+
+        #oh god oh fuck
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv7.py]: >:|

File: bemani/backend/dm/dmv8.py [hunk 19]
```diff
@@ -0,0 +1,2503 @@
...
+
+class DrummaniaV8(
+    EventLogHandler,
+    DrummaniaBase,
+):
+
+    name = "Drummania V8"
+    version = VersionConstants.DRUMMANIA_V8
+
+    GAME_LIMITED_LOCKED = 1
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv8.py:23]: I assume that GFDMv8 is working? Why is there so much SDVX copypasta here still? Did you run lint on this even one time?

File: bemani/backend/dm/dmv8.py [hunk 20]
```diff
@@ -0,0 +1,2503 @@
...
+    GAME_GRADE_AA = 5
+    GAME_GRADE_AAA = 6
+    GAME_GRADE_S = 7
+
+    @classmethod
+
+    def previous_version(self) -> Optional[DrummaniaBase]:
+        return DrummaniaV7(self.data, self.config, self.model)
+
+    def db_to_game_chart(self, db_chart: int) -> int:
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv8.py:52]: Is there a corresponding game_to_db_chart? I don't see any of the constants being used above. How do you convert between game constants and DB constants? Am I missing something?

File: bemani/backend/dm/dmv8.py [hunk 21]
```diff
@@ -0,0 +1,2503 @@
...
+        bossdata.add_child(Node.u8('adv_encore_border', 91))
+        bossdata.add_child(Node.u8('ext_encore_border', 92))
+        bossdata.add_child(Node.u8('bsc_premium_border', 93))
+        bossdata.add_child(Node.u8('adv_premium_border', 93))
+        bossdata.add_child(Node.u8('ext_premium_border', 93))
+
+        info = Node.void('info')
+        root.add_child(info)
+
+        info.add_child(Node.string('message', "Brought to you by PhaseII! Hosted on phaseii.iidxfan.xyz. Login today!"))
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv8.py]: Yeah no, what the fuck? Did you check or test any of this code before putting up a PR?

File: bemani/backend/dm/dmv8.py [hunk 22]
```diff
@@ -0,0 +1,2503 @@
...
+
+        return root
+
+    def handle_cardutil_regist_request(self, request: Node) -> Node:
+
+        data = request.child('data')
+
+        refid = data.child_value('refid')
+        name = data.child_value('name')
+        gdp = 0
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv8.py:694]: Same comment here, just move these into the dict below??

File: bemani/backend/dm/dmv8.py [hunk 23]
```diff
@@ -0,0 +1,2503 @@
...
+        battledata.add_child(Node.u32('perfect', profile.get_int('perfect', 0)))
+        battledata.add_child(Node.u32('great', profile.get_int('great', 0)))
+        battledata.add_child(Node.u32('good', profile.get_int('good', 0)))
+        battledata.add_child(Node.u32('poor', profile.get_int('poor', 0)))
+        battledata.add_child(Node.u32('miss', profile.get_int('miss', 0)))
+
+        history = Node.void('history')
+        battledata.add_child(history)
+
+        round_before = Node.void('round')
```
Comment (DragonMinded) [inline bemani/backend/dm/dmv8.py:945]: None of this is filled in? Does DMv8 work? Is it considered finished? Why is there copypasta for miles instead of a dang loop?

File: bemani/backend/dm/base.py [hunk 24]
```diff
@@ -0,0 +1,354 @@
...
+        return self.version
+
+    def previous_version(self) -> Optional['DrummaniaBase']:
+        """
+        Returns the previous version of the game, based on this game. Should
+        be overridden.
+        """
+        return None
+    
+    def get_profile_by_refid(self, refid: Optional[str]) -> Optional[Node]:
```
Comment (DragonMinded) [inline bemani/backend/dm/base.py:55]: It seems that this, new_profile_by_refid, put_profile_by_refid and get_clear_rates are all completely unused. Why are they included in this file?

File: bemani/backend/gf/base.py [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+
+from bemani.backend.base import Base
+from bemani.backend.core import CoreHandler, CardManagerHandler
+from bemani.common import ValidatedDict, GameConstants, DBConstants, Parallel, Model
+from bemani.data import UserID, Score, Data
+from bemani.protocol import Node
+
+class GuitarFreaksBase(CoreHandler, CardManagerHandler, Base):
+    """
+    Trying to work on GFDM right now. now it's time to play the guitar rhythmically :)
```
Comment (DragonMinded) [inline bemani/backend/gf/base.py:11]: Again, please don't pollute docstrings like this.

File: bemani/backend/gf/base.py [hunk 26]
```diff
@@ -0,0 +1,351 @@
...
+        #write our skill percent
+        scoredata.replace_int('skill_perc', max(scoredata.get_int('skill_perc'), skill_perc))
+        history.replace_int('skill_perc', skill_perc)
+
+        #did we clear the song?
+        scoredata.replace_int('clear', max(scoredata.get_int('clear'), clear))
+        history.replace_int('clear', clear)
+
+        #did we cheat?
+        #scoredata.replace_int('autoclear', max(scoredata.get_int('autoclear'), autoclear))
```
Comment (DragonMinded) [inline bemani/backend/gf/base.py:275]: Same comment as DM here.

File: bemani/backend/gf/base.py [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+        print(songid)
+
+        # GFDM for all versions sends all of the songs
+        # a player played at the end of the round. It doesn't send timestamps
+        # for those songs (Jubeat does). So, if a user plays the same song/chart
+        # more than once in a round, we will end up failing to store the attempt
+        # since we don't allow two of the same attempt at the same time for the
+        # same user and song/chart. So, bump the timestamp by one second and retry
+        # well past the maximum number of songs.
+        now = Time.now()
```
Comment (DragonMinded) [inline bemani/backend/gf/base.py:311]: Also same comment, this code does literally nothing other than change every single timestamp by 55 seconds.

File: bemani/backend/gf/gfv8.py [hunk 28]
```diff
@@ -0,0 +1,2657 @@
...
+    GuitarFreaksBase,
+):
+
+    name = "GuitarFreaks V8"
+    version = VersionConstants.GUITARFREAKS_V8
+
+    GAME_LIMITED_LOCKED = 1
+    GAME_LIMITED_UNLOCKED = 2
+
+    GAME_CURRENCY_PACKETS = 0
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:26]: Looks like SDVX copypasta here too? I skipped reviewing gfv4-v7 because I assume you did the same thing with them as you did with DM.

File: bemani/backend/gf/gfv8.py [hunk 29]
```diff
@@ -0,0 +1,2657 @@
...
+    GAME_GRADE_AA = 5
+    GAME_GRADE_AAA = 6
+    GAME_GRADE_S = 7
+
+    @classmethod
+
+    def previous_version(self) -> Optional[GuitarFreaksBase]:
+        return GuitarFreaksV7(self.data, self.config, self.model)
+
+    def db_to_game_chart(self, db_chart: int) -> int:
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:60]: Where's game_to_db_chart? Where is this used?

File: bemani/backend/gf/gfv8.py [hunk 30]
```diff
@@ -0,0 +1,2657 @@
...
+            self.CHART_TYPE_BASS_BASIC: self.GAME_CHART_TYPE_BASS_BASIC,
+            self.CHART_TYPE_BASS_ADVANCED: self.GAME_CHART_TYPE_BASS_ADVANCED,
+            self.CHART_TYPE_BASS_EXTREME: self.GAME_CHART_TYPE_BASS_EXTREME,
+            self.CHART_TYPE_OPEN_BEGINNER: self.GAME_CHART_TYPE_OPEN_BEGINNER,
+            self.CHART_TYPE_OPEN_BASIC: self.GAME_CHART_TYPE_OPEN_BASIC,
+            self.CHART_TYPE_OPEN_ADVANCED: self.GAME_CHART_TYPE_OPEN_ADVANCED,
+            self.CHART_TYPE_OPEN_EXTREME: self.GAME_CHART_TYPE_OPEN_EXTREME,
+        }[db_chart]
+
+    def handle_shopinfo_regist_request(self, request: Node) -> Node:
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:76]: Is this used?

File: bemani/backend/gf/gfv8.py [hunk 31]
```diff
@@ -0,0 +1,2657 @@
...
+        bossdata.add_child(Node.u8('adv_encore_border', 91))
+        bossdata.add_child(Node.u8('ext_encore_border', 92))
+        bossdata.add_child(Node.u8('bsc_premium_border', 93))
+        bossdata.add_child(Node.u8('adv_premium_border', 93))
+        bossdata.add_child(Node.u8('ext_premium_border', 93))
+
+        info = Node.void('info')
+        root.add_child(info)
+
+        info.add_child(Node.string('message', "Brought to you by PhaseII! Hosted on phaseii.iidxfan.xyz. Login today!"))
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py]: Dude really? This is about as bad as bundle sites advertising in the zip file where you downloaded somebody else's software.

File: bemani/backend/gf/gfv8.py [hunk 32]
```diff
@@ -0,0 +1,2657 @@
...
+            'is_succession': is_succession,
+            'lastmode': lastmode,
+        })
+
+        self.put_profile(userid, defaultprofile)
+
+        # Now, reload and format the profile, looking up the has old version flag
+        profile = self.get_profile(userid)
+
+        # oldversion = self.previous_version()
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:748]: Why is this commented out?

File: bemani/backend/gf/gfv8.py [hunk 33]
```diff
@@ -0,0 +1,2657 @@
...
+        battledata.add_child(Node.u32('perfect', profile.get_int('perfect', 0)))
+        battledata.add_child(Node.u32('great', profile.get_int('great', 0)))
+        battledata.add_child(Node.u32('good', profile.get_int('good', 0)))
+        battledata.add_child(Node.u32('poor', profile.get_int('poor', 0)))
+        battledata.add_child(Node.u32('miss', profile.get_int('miss', 0)))
+
+        history = Node.void('history')
+        battledata.add_child(history)
+
+        round_before = Node.void('round')
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:972]: Again with the completely unpopulated chunks, why is this not handled and not in a loop? You have access to the score history, this doesn't look done at all.

File: bemani/backend/gf/gfv8.py [hunk 34]
```diff
@@ -0,0 +1,2657 @@
...
+    
+    def unformat_profile(self, userid: UserID, request: Node, oldprofile: ValidatedDict) -> ValidatedDict:
+        newprofile = copy.deepcopy(oldprofile)
+
+        # get base song info
+        modedata = request.child('modedata')
+        playedmode = modedata.attribute('mode')
+
+        if playedmode == "standard" :
+            lastmode = 1
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:2263]: Use defines instead of magic values here.

File: bemani/backend/gf/gfv8.py [hunk 35]
```diff
@@ -0,0 +1,2657 @@
...
+                self.update_score(userid, songid, seqmode, clear, autoclear, score, fullcombo, excellent, combo, skill_point, skill_perc, result_rank, combo_rate, perfect_rate, difficulty)
+
+        return newprofile
+
+    def format_scores(self, userid: UserID, profile: ValidatedDict, scores: List[Score]) -> Node:
+        scores = self.data.remote.music.get_scores(self.game, self.music_version, userid)
+
+        standard = Node.void('standard')
+
+        easy = 0
```
Comment (DragonMinded) [inline bemani/backend/gf/gfv8.py:2445]: Aren't these exactly the same as the constants above? Why are these here?

File: bemani/common/constants.py [hunk 36]
```diff
@@ -1,4 +1,5 @@
 class GameConstants:
+    ROAD_FIGHTERS = 'roadfighters'
```
Comment (DragonMinded) [inline bemani/common/constants.py:2]: Why is this here? This is not related to the PR.

File: bemani/common/constants.py [hunk 37]
```diff
@@ -8,11 +9,39 @@ class GameConstants:
     POPN_MUSIC = 'pnm'
     REFLEC_BEAT = 'reflec'
     SDVX = 'sdvx'
+    DRUMMANIA = 'dm'
+    GUITARFREAKS = 'gf'
+    GITADORA = 'gtd'
+    POPN_HELLO = 'hpm'
+    BEATSTREAM = 'bst'
```
Comment (DragonMinded) [inline bemani/common/constants.py:16]: Same with all of these. Please only include relevant changes.

File: bemani/common/constants.py [hunk 38]
```diff
@@ -164,6 +250,28 @@ class DBConstants:
     DDR_RANK_AA_PLUS = 650
     DDR_RANK_AAA = 700
 
+    DANCE_EVOLUTION_HALO_NONE = 100
```
Comment (DragonMinded) [inline bemani/common/constants.py:253]: Same with all of this.

File: bemani/utils/services.py [hunk 39]
```diff
@@ -9,12 +9,19 @@
...
 from bemani.backend.iidx import IIDXFactory
 from bemani.backend.popn import PopnMusicFactory
+from bemani.backend.popnhell import PopnHelloFactory
 from bemani.backend.jubeat import JubeatFactory
 from bemani.backend.bishi import BishiBashiFactory
 from bemani.backend.ddr import DDRFactory
 from bemani.backend.sdvx import SoundVoltexFactory
 from bemani.backend.reflec import ReflecBeatFactory
 from bemani.backend.museca import MusecaFactory
+#from bemani.backend.roadfighters import RoadFightersFactory
```
Comment (DragonMinded) [inline bemani/utils/services.py:19]: Get rid of this commented out stuff.

File: bemani/utils/services.py [hunk 40]
```diff
@@ -151,6 +160,18 @@ def register_games() -> None:
         SoundVoltexFactory.register_all()
     if config.get('support', {}).get(GameConstants.MUSECA, False):
         MusecaFactory.register_all()
+    #if config.get('support', {}).get(GameConstants.ROAD_FIGHTERS, False):
```
Comment (DragonMinded) [inline bemani/utils/services.py:163]: Same here.

File: bemani/backend/gf/base.py [hunk 41]
```diff
@@ -0,0 +1,351 @@
...
+            'name': name,
+            'loc': locid,
+        })
+        self.put_profile(userid, defaultprofile)
+
+        # Now, reload and format the profile, looking up the has old version flag
+        profile = self.get_profile(userid)
+        return self.format_profile(userid, profile)
+
+    def put_profile_by_refid(self, refid: Optional[str], request: Node) -> None:
```
Comment (DragonMinded) [inline bemani/backend/gf/base.py:91]: Same comment here for all _by_refid functions, they appear completely unused. Is this true?

File: bemani/backend/popnhell/base.py [hunk 42]
```diff
@@ -0,0 +1,98 @@
...
+    game = GameConstants.POPN_HELLO
+
+    def previous_version(self) -> Optional['PopnHelloBase']:
+        """
+        Returns the previous version of the game, based on this game. Should
+        be overridden.
+        """
+        return None
+
+    def new_profile_by_refid(self, refid: Optional[str], name: Optional[str], locid: Optional[int]) -> Node:
```
Comment (DragonMinded) [inline bemani/backend/popnhell/base.py:27]: Same here, these don't appear to be used by the implementation at all.

File: bemani/backend/popnhell/popnhello.py [hunk 43]
```diff
@@ -0,0 +1,137 @@
...
+
+    @classmethod
+    def handle_game_common_request(self, request: Node) -> Node:
+        #game_common
+        root = Node.void('game')
+
+        flag = Node.void('flag')
+        root.add_child(flag)
+
+        flag.set_attribute(id, '1')
```
Comment (DragonMinded) [inline bemani/backend/popnhell/popnhello.py]: Was this even tested? None of these variables are defined, this does nothing but crash?

File: bemani/backend/popnhell/popnhello.py [hunk 44]
```diff
@@ -0,0 +1,137 @@
...
+    def handle_game_shop_request(self, request: Node) -> Node:
+        #game_shop
+        root = Node.void('game')
+
+        return root
+
+    def handle_game_new_request(self, request: Node) -> Node:
+        #game_new
+        root = Node.void('game')
+        name = "0"
```
Comment (DragonMinded) [inline bemani/backend/popnhell/popnhello.py]: Why are these here, and why not in the validated dict? Why do they even have to be in the default profile at all? get_int will return a "0" for anything not in the profile already, this is entirely useless. Also, I realize that that's the case where you did this for gfdm as well. Best to remove it.

File: bemani/backend/popnhell/popnhello.py [hunk 45]
```diff
@@ -0,0 +1,137 @@
...
+        root.add_child(last)
+        last.set_attribute('chara', profile.get_str('chara'))
+        last.set_attribute('level', profile.get_str('level'))
+        last.set_attribute('music_id', profile.get_str('last_music'))
+        last.set_attribute('style', profile.get_str('style'))
+
+
+        return root
+
+    def handle_game_load_m_request(self, request: Node) -> Node:
```
Comment (DragonMinded) [inline bemani/backend/popnhell/popnhello.py:90]: Is this in hello popn? Where are scores loaded and saved? This looks like it boots the game and that's it.

File: bemani/utils/read.py [hunk 46]
```diff
@@ -3482,23 +5383,24 @@ def import_music_db(self, songs: List[Dict[str, Any]]) -> None:
...
+            iidx.import_qpros(qpros)
         iidx.close()
 
     elif args.series == GameConstants.DDR:
         ddr = ImportDDR(config, args.version, args.no_combine, args.update)
         if args.server and args.token:
             songs = ddr.lookup(args.server, args.token)
         else:
-            if args.version == '16':
+            if args.version >= '16':
```
Comment (DragonMinded) [inline bemani/utils/read.py:5228]: Unrelated change?

File: bemani/utils/read.py [hunk 47]
```diff
@@ -3482,23 +5383,24 @@ def import_music_db(self, songs: List[Dict[str, Any]]) -> None:
...
             elif args.server and args.token:
-                songs = iidx.lookup(args.server, args.token)
+                songs, qpros = iidx.lookup(args.server, args.token)
             else:
                 raise Exception(
                     'No music_data.bin or TSV provided and no remote server specified! Please ' +
                     'provide either a --bin, --tsv or a --server and --token option!'
                 )
             iidx.import_music_db(songs)
+            iidx.import_qpros(qpros)
```
Comment (DragonMinded) [inline bemani/utils/read.py:5220]: Unrelated?

File: bemani/utils/read.py [hunk 48]
```diff
@@ -3473,6 +5313,67 @@ def import_music_db(self, songs: List[Dict[str, Any]]) -> None:
...
+                    'No music_info.xml or TSV provided and no remote server specified! Please ' +
+                    'provide either a --xml, --tsv or a --server and --token option!'
+                )
+            dm.import_music_db(songs)
+        dm.close()
+
+    elif args.series == GameConstants.GUITARFREAKS:
+        gf = ImportGuitarFreaks(config, args.version, args.no_combine, args.update)
+        if args.tsv is not None:
+            # Special case for Jubeat, grab the title/artist metadata that was
```
Comment (DragonMinded) [inline bemani/utils/read.py]: Copypasta?

File: bemani/utils/read.py [hunk 49]
```diff
@@ -3473,6 +5313,67 @@ def import_music_db(self, songs: List[Dict[str, Any]]) -> None:
...
+            # hand-populated since its not in the music DB.
+            gf.import_metadata(args.tsv)
+        else:
+            # Normal case, doing a music DB or emblem import.
```
Comment (DragonMinded) [inline bemani/utils/read.py:5173]: This looks completely copypasta, and there's no tsv for gfdm songs, why is a lot of this even here?

File: bemani/utils/read.py [hunk 50]
```diff
@@ -12,7 +12,7 @@
 import yaml  # type: ignore
 import xml.etree.ElementTree as ET
 from sqlalchemy import create_engine  # type: ignore
-from sqlalchemy.engine import CursorResult  # type: ignore
+from sqlalchemy.engine.result import ResultProxy  # type: ignore
```
Comment (DragonMinded) [inline bemani/utils/read.py:15]: This reverts the change for sqlalchemy compatibility?

**Review by DragonMinded — CHANGES_REQUESTED** [review]
Thanks for putting this up! However, this is so far from mergeable that I suspect you'll have to spend a lot of time working on it. There's several files where you made them executable for some reason, read.py is so littered with unrelated changes that I gave up reviewing it, there's a massive amount of copypasta, there seems to be so many lint errors and I see absolutely no tests whatsoever which would have caught some bugs in the code. None of the stuff that's listed as working is complete. You also are trying to advertise your own server instance in the code which is tacky at best.

File: bemani/backend/dm/base.py (hunk 7 above)
Comment (trmazi) [reply bemani/backend/dm/base.py:280]: the game sends that if you use auto pedal
it doesn't read it anywhere though, so i chose not to record it

File: bemani/backend/dm/base.py (hunk 8 above)
Comment (trmazi) [reply bemani/backend/dm/base.py:315]: the game does not send a timestamp :(
if there's a cleaner way to do this, please let me know

File: bemani/backend/dm/base.py (hunk 9 above)
Comment (trmazi) [reply bemani/backend/dm/base.py:322]: it was tested, and works most of the time (not good)

File: bemani/backend/dm/dmv4.py (hunk 10 above)
Comment (trmazi) [reply bemani/backend/dm/dmv4.py]: it's copy pasta, yes. forgot to remove this from the public fork..... oops

File: bemani/backend/dm/dmv5.py (hunk 11 above)
Comment (trmazi) [reply bemani/backend/dm/dmv5.py]: it's copy pasta, yes. forgot to remove this from the public fork..... oops

File: bemani/backend/dm/dmv6.py (hunk 12 above)
Comment (trmazi) [reply bemani/backend/dm/dmv6.py]: it's copy pasta, yes. forgot to remove this from the public fork..... oops

File: bemani/backend/dm/dmv7.py (hunk 13 above)
Comment (trmazi) [reply bemani/backend/dm/dmv7.py]: shouldn't be copypasta?? but maybe sent the wrong thing

File: bemani/backend/dm/dmv7.py (hunk 14 above)
Comment (trmazi) [reply bemani/backend/dm/dmv7.py]: Was cleaned in a later version

File: bemani/backend/dm/dmv7.py (hunk 16 above)
Comment (trmazi) [reply bemani/backend/dm/dmv7.py]: loops are hard, this was also like one of the first things i wrote. fixing this soon though.

File: bemani/backend/dm/dmv7.py (hunk 17 above)
Comment (trmazi) [reply bemani/backend/dm/dmv7.py]: because my dumb ass sent the wrong file. this is from before i had any idea on how to load the scores

File: bemani/backend/dm/dmv8.py (hunk 21 above)
Comment (trmazi) [reply bemani/backend/dm/dmv8.py]: i DEF forgot to remove that, sorry. 
this is going to be replaced with an arcade webui setting soon.

File: bemani/backend/dm/dmv8.py (hunk 22 above)
Comment (trmazi) [reply bemani/backend/dm/dmv8.py:694]: same response above.

File: bemani/backend/dm/base.py (hunk 24 above)
Comment (trmazi) [reply bemani/backend/dm/base.py:55]: because i forgot to send the right file

File: bemani/backend/popnhell/popnhello.py (hunk 43 above)
Comment (trmazi) [reply bemani/backend/popnhell/popnhello.py]: sent the wrong file. this will def break things.

File: bemani/utils/read.py (hunk 48 above)
Comment (trmazi) [reply bemani/utils/read.py]: GFDM also needs a tsv for artist and song data.
forgot to change that 😨

File: bemani/utils/read.py (hunk 49 above)
Comment (trmazi) [reply bemani/utils/read.py:5173]: again, GFDM requires a TSV, like jubeat/iidx

File: bemani/backend/dm/base.py (hunk 7 above)
Comment (DragonMinded) [reply bemani/backend/dm/base.py:280]: Okay, then we should probably get rid of this section as well as the autoclear parameter.

File: bemani/backend/dm/base.py (hunk 8 above)
Comment (DragonMinded) [reply bemani/backend/dm/base.py:315]: Several games don't send timestamps. That's not what this comment is pointing out. This comment is pointing out that you are literally just adding in a loop. You don't actually handle multiple attempts of the same song.

File: bemani/backend/dm/base.py (hunk 9 above)
Comment (DragonMinded) [reply bemani/backend/dm/base.py:322]: Yeah, this code doesn't do what you think it does, it is literally not trying to insert scores and fixing duplicates at all.

**Comment by DragonMinded** [comment]
This is clearly going nowhere and was a complete waste of my time to even review. Closing because it is cluttering PRs.
